kv/kvserver: TestReplicaCircuitBreaker_Partial_Retry failed

kv/kvserver.TestReplicaCircuitBreaker_Partial_Retry [failed](https://tanzanite.cluster.engflow.com/invocations/default/7d49445a-9d67-46fc-9f66-44f379fc2719?testReportRun=7&testReportShard=24&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on release-24.1 @ [d9faf8448f56be84ba5cafdda1bc8b0a3158a9e7](https://github.com/cockroachdb/cockroach/commits/d9faf8448f56be84ba5cafdda1bc8b0a3158a9e7):


```
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/client_replica_circuit_breaker_test.go:807
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/client_replica_circuit_breaker_test.go:870
        	Error:      	Received unexpected error:
        	            	replica unavailable: (n3,s3):3 unable to serve request to r69:/{Table/Max-Max} [(n1,s1):1, (n2,s2):2, (n3,s3):3, next=4, gen=5, sticky=9223372036.854775807,2147483647]: closed timestamp: 1730269992.146873138,0 (2024-10-30 06:33:12); raft status: {"id":"3","term":7,"vote":"3","commit":26,"lead":"3","raftState":"StateLeader","applied":26,"progress":{"3":{"match":26,"next":27,"state":"StateReplicate"},"1":{"match":26,"next":27,"state":"StateReplicate"},"2":{"match":26,"next":27,"state":"StateReplicate"}},"leadtransferee":"0"}: injected error
        	            	(1) forced error mark
        	            	  | "originated at breaker breaker"
        	            	  | github.com/cockroachdb/cockroach/pkg/util/circuit/*circuit.breakerErrorMark::
        	            	Wraps: (2) forced error mark
        	            	  | "breaker open"
        	            	  | github.com/cockroachdb/errors/withstack/*withstack.withStack::
        	            	Wraps: (3) replica unavailable: (n3,s3):3 unable to serve request to r69:/{Table/Max-Max} [(n1,s1):1, (n2,s2):2, (n3,s3):3, next=4, gen=5, sticky=9223372036.854775807,2147483647]: closed timestamp: 1730269992.146873138,0 (2024-10-30 06:33:12); raft status: {"id":"3","term":7,"vote":"3","commit":26,"lead":"3","raftState":"StateLeader","applied":26,"progress":{"3":{"match":26,"next":27,"state":"StateReplicate"},"1":{"match":26,"next":27,"state":"StateReplicate"},"2":{"match":26,"next":27,"state":"StateReplicate"}},"leadtransferee":"0"}: injected error
        	            	Wraps: (4)
        	            	  | (opaque error wrapper)
        	            	  | type name: github.com/cockroachdb/errors/withstack/*withstack.withStack
        	            	  | reportable 0:
        	            	  |
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.replicaUnavailableError
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_circuit_breaker.go:268
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).replicaUnavailableError
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_circuit_breaker.go:277
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*replicaCircuitBreaker).tripSync
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_circuit_breaker.go:97
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).TripBreaker
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/helpers_test.go:597
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestReplicaCircuitBreaker_Partial_Retry
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/client_replica_circuit_breaker_test.go:838
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1689
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1695
        	            	Wraps: (5) closed timestamp: 1730269992.146873138,0 (2024-10-30 06:33:12); raft status: {"id":"3","term":7,"vote":"3","commit":26,"lead":"3","raftState":"StateLeader","applied":26,"progress":{"3":{"match":26,"next":27,"state":"StateReplicate"},"1":{"match":26,"next":27,"state":"StateReplicate"},"2":{"match":26,"next":27,"state":"StateReplicate"}},"leadtransferee":"0"}
        	            	Wraps: (6)
        	            	  | (opaque error wrapper)
        	            	  | type name: github.com/cockroachdb/errors/withstack/*withstack.withStack
        	            	  | reportable 0:
        	            	  |
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).TripBreaker
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/helpers_test.go:597
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestReplicaCircuitBreaker_Partial_Retry
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/client_replica_circuit_breaker_test.go:838
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1689
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1695
        	            	Wraps: (7) injected error
        	            	Error types: (1) *markers.withMark (2) *markers.withMark (3) *kvpb.ReplicaUnavailableError (4) *errbase.opaqueWrapper (5) *errutil.withPrefix (6) *errbase.opaqueWrapper (7) *errutil.leafError
        	Test:       	TestReplicaCircuitBreaker_Partial_Retry
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestReplicaCircuitBreaker_Partial_Retry205057717
--- FAIL: TestReplicaCircuitBreaker_Partial_Retry (5.38s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=7</code>
 - <code>shard=24</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReplicaCircuitBreaker_Partial_Retry.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-43779

--- a/pkg/kv/kvserver/client_replica_circuit_breaker_test.go
+++ b/pkg/kv/kvserver/client_replica_circuit_breaker_test.go
@@ -883,9 +883,16 @@ func TestReplicaCircuitBreaker_Partial_Retry(t *testing.T) {
 	var leader raftpb.PeerID
 	require.Eventually(t, func() bool {
 		for _, repl := range repls {
-			if l := repl.RaftStatus().Lead; l == 3 {
+			l := repl.RaftStatus().Lead
+			if l == 3 {
 				return false
-			} else if l > 0 {
+			}
+			if repl.ReplicaID() != 3 && l == 0 {
+				// The old leader (3) steps down because of check quorum. A new leader
+				// should be elected amongst 1 and 2, and both of them should know who
+				// that is before we proceed with the test.
+				return false
+			} else if repl.ReplicaID() != 3 {
 				leader = l
 			}
 		}

